### PR TITLE
[Feature](Row store) support column group with store row format for partial columns of table

### PR DESCRIPTION
**Pull request by eldenmoon**
## Proposed changes
Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/analysis/CreateTableStmt.java [hunk 1]
```diff
@@ -468,6 +469,36 @@ public void analyze(Analyzer analyzer) throws UserException {
                 columnDefs.add(ColumnDef.newVersionColumnDef(AggregateType.REPLACE));
             }
         }
+
+        if (properties != null) {
```
Comment (xiaokang) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/CreateTableStmt.java:473]: put it inside row store branch

File: fe/fe-core/src/main/java/org/apache/doris/analysis/CreateTableStmt.java [hunk 2]
```diff
@@ -468,6 +469,36 @@ public void analyze(Analyzer analyzer) throws UserException {
...
+            Map<String, List<String>> columnGroups = PropertyAnalyzer.analyzeColumnGroups(Maps.newHashMap(properties));
```
Comment (xiaokang) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/CreateTableStmt.java:474]: It's duplicate with CreateTableInfo.

**Review by xiaokang — COMMENTED** [review]
fe reviewed

File: be/src/service/point_query_executor.h [hunk 3]
```diff
@@ -103,6 +111,12 @@ class Reusable {
     vectorized::DataTypeSerDeSPtrs _data_type_serdes;
     std::unordered_map<uint32_t, uint32_t> _col_uid_to_idx;
     std::vector<std::string> _col_default_values;
+    // picked rowstore(column group) column unique id
+    int32_t _rs_column_uid;
```
Comment (xiaokang) [inline be/src/service/point_query_executor.h:115]: what's the meaning of name rs?

File: be/src/service/point_query_executor.cpp [hunk 4]
```diff
@@ -47,16 +52,103 @@
...
+        missing_cids.insert(slot->col_unique_id());
+    }
+    if (target_rs_column_id == -1) {
+        // no row store columns
+        return;
+    }
+    const TabletColumn& target_rs_column = schema.column_by_uid(target_rs_column_id);
+    DCHECK(target_rs_column.is_row_store_column());
+    // The full column group is considered a full match, thus no missing cids
+    if (target_rs_column.group_ids().empty()) {
```
Comment (xiaokang) [inline be/src/service/point_query_executor.cpp:124]: consider that if it's right after light schema chage

File: be/src/olap/base_tablet.cpp [hunk 5]
```diff
@@ -909,7 +908,7 @@ Status BaseTablet::fetch_value_through_row_column(RowsetSharedPtr input_rowset,
         serdes[i] = type->get_serde();
     }
     vectorized::JsonbSerializeUtil::jsonb_to_block(serdes, *string_column, col_uid_to_idx, block,
-                                                   default_values);
+                                                   default_values, {});
```
Comment (xiaokang) [inline be/src/olap/base_tablet.cpp:911]: Do you mean use all columns from row column? Can it be optimized to only copy needed columns?

File: be/src/exec/rowid_fetcher.cpp [hunk 6]
```diff
@@ -405,12 +405,14 @@ Status RowIdStorageReader::read_by_rowids(const PMultiGetRequest& request,
...
-            CHECK(tablet->tablet_schema()->store_row_column());
+            CHECK(tablet->tablet_schema()->has_full_row_store_column());
             RowLocation loc(rowset_id, segment->id(), row_loc.ordinal_id());
             string* value = response->add_binary_row_data();
             RETURN_IF_ERROR(scope_timer_run(
                     [&]() {
-                        return tablet->lookup_row_data({}, loc, rowset, &desc, stats, *value);
+                        return tablet->lookup_row_data(
+                                {}, loc, rowset, &desc, stats, *value,
+                                tablet->tablet_schema()->column(BeConsts::ROW_STORE_COL));
```
Comment (xiaokang) [inline be/src/exec/rowid_fetcher.cpp:415]: Why not specific row column xxx?

**Review by xiaokang — COMMENTED** [review]
be reviewed

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

File: be/src/service/point_query_executor.cpp [hunk 7]
```diff
@@ -80,6 +131,23 @@ Status Reusable::init(const TDescriptorTable& t_desc_tbl, const std::vector<TExp
...
+    // Get the output slot descriptors
+    std::vector<SlotDescriptor*> output_slot_descs;
+    std::vector<vectorized::VExprSPtr> slot_exprs;
+    for (const auto& expr : _output_exprs_ctxs) {
+        extract_slot_ref(expr->root(), slot_exprs);
+    }
+    for (const auto& expr : slot_exprs) {
+        int column_id = static_cast<const vectorized::VSlotRef*>(expr.get())->column_id();
+        auto* slot_desc = tuple_desc()->slots()[column_id];
+        output_slot_descs.push_back(slot_desc);
```
Comment (xiaokang) [inline be/src/service/point_query_executor.cpp:144]: You can change `extract_slot_ref` to `extract_slot_desc` and extract `output_slot_descs` in it directlly.

File: be/src/service/point_query_executor.cpp [hunk 8]
```diff
@@ -335,23 +407,68 @@ Status PointQueryExecutor::_lookup_row_data() {
                     _reusable->get_data_type_serdes(),
                     _row_read_ctxs[i]._cached_row_data.data().data,
                     _row_read_ctxs[i]._cached_row_data.data().size, _reusable->get_col_uid_to_idx(),
-                    *_result_block, _reusable->get_col_default_values());
+                    *_result_block, _reusable->get_col_default_values(),
+                    _reusable->include_col_uids());
             continue;
```
Comment (xiaokang) [inline be/src/service/point_query_executor.cpp:412]: Do you need to consider missing columns?

**Review by xiaokang — COMMENTED** [review]
suggest a consistent name row_store

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/info/CreateMTMVInfo.java [hunk 9]
```diff
@@ -340,7 +340,7 @@ private void getColumns(Plan plan) {
         if (properties != null) {
             try {
                 boolean storeRowColumn =
-                        PropertyAnalyzer.analyzeStoreRowColumn(Maps.newHashMap(properties));
+                        PropertyAnalyzer.analyzeStoreRowColumn(Maps.newHashMap(properties), null, true);
```
[the remark this answers is not in the document]
Comment (xiaokang) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/info/CreateMTMVInfo.java:343]: add support for partial

File: be/src/vec/jsonb/serialize.cpp [hunk 10]
```diff
@@ -55,12 +57,15 @@ void JsonbSerializeUtil::block_to_jsonb(const TabletSchema& schema, const Block&
...
             const auto& column = block.get_by_position(j).column;
             const auto& tablet_column = *schema.columns()[j];
+            // ignore row store columns
             if (tablet_column.is_row_store_column()) {
-                // ignore dst row store column
                 continue;
             }
-            serdes[j]->write_one_cell_to_jsonb(*column, jsonb_writer, &pool,
-                                               tablet_column.unique_id(), i);
+            // TODO improve performance for checking column in group
```
[the remark this answers is not in the document]
Comment (xiaokang) [reply be/src/vec/jsonb/serialize.cpp:64]: preprocess a vector<bool>

File: be/src/service/point_query_executor.h (hunk 3 above)
Comment (xiaokang) [reply be/src/service/point_query_executor.h:115]: _row_store_coloumn_uid

File: be/src/common/config.h [hunk 11]
```diff
@@ -1370,6 +1370,9 @@ DECLARE_Bool(enable_jvm_monitor);
 // Skip loading stale rowset meta when initializing `TabletMeta` from protobuf
 DECLARE_mBool(skip_loading_stale_rowset_meta);
 
+// Whether to enable access column store in short circuit quries, otherwise report error
+DECLARE_mBool(enable_short_circuit_query_access_column_store);
```
[the remark this answers is not in the document]
Comment (xiaokang) [reply be/src/common/config.h:1374]: change to session var

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

File: be/src/service/point_query_executor.cpp [hunk 12]
```diff
@@ -335,23 +406,63 @@ Status PointQueryExecutor::_lookup_row_data() {
...
-                _profile_metrics.read_stats, value,
-                !config::disable_storage_row_cache /*whether write row cache*/));
-        // serilize value to block, currently only jsonb row formt
-        vectorized::JsonbSerializeUtil::jsonb_to_block(
-                _reusable->get_data_type_serdes(), value.data(), value.size(),
-                _reusable->get_col_uid_to_idx(), *_result_block,
-                _reusable->get_col_default_values());
+        // fill block by row store
+        if (_reusable->rs_column_uid() != -1) {
+            bool use_row_cache = !config::disable_storage_row_cache &&
```
[the remark this answers is not in the document]
Comment (xiaokang) [reply be/src/service/point_query_executor.cpp:419]: What's the problem if row cache is used for partial row store? I think it's still OK.

File: fe/fe-core/src/main/java/org/apache/doris/alter/SchemaChangeHandler.java [hunk 13]
```diff
@@ -1920,6 +1946,20 @@ public int getAsInt() {
                         }
                         enableLightSchemaChange(db, olapTable);
                         return;
+                    } else if (properties.containsKey(PropertyAnalyzer.PROPERTIES_STORE_ROW_COLUMN)
+                                || properties.containsKey(PropertyAnalyzer.PROPERTIES_ROW_STORE_COLUMNS)) {
+                        String value = properties.get(PropertyAnalyzer.PROPERTIES_STORE_ROW_COLUMN);
+                        if (value != null && value.equalsIgnoreCase("false")) {
+                            throw new DdlException("Can not alter store_row_column from true to false currently");
```
Comment (xiaokang) [inline fe/fe-core/src/main/java/org/apache/doris/alter/SchemaChangeHandler.java:1953]: User may want to drop row store since it's space consuming.

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

**Review by xiaokang — APPROVED** [review]
LGTM

**Review by lxr599 — APPROVED** [review]

**Pull request merged**